### PR TITLE
Fix guess mix with SAD and other guesses

### PR DESCRIPTION
**Pull request by susilehtola**
## Description

The current guess mix code runs in iteration 0, but SAD doesn't have orbitals there, while some other guesses don't call `form_C` in iteration 0. This PR fixes guess_mix for all guesses.

## Todos

- [x] Tests for guess_mix added for all guesses

## Questions
- [ ] Question1

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

**Comment by JonathonMisiewicz** [comment]
The PR mostly looks good. A few questions:
1. Can we get an `output.ref` on the test?
2. On the linked issue (#2410), you said this is a "tentative" fix. Does that mean anything besides that the PR isn't accepted yet?
3. I'll hold off on accepting until the "documentation" issue is settled.

**Comment by susilehtola** [comment]
>     2. On the linked issue ([Wrong guess_mix when guess=sad #2410](https://github.com/psi4/psi4/issues/2410)), you said this is a "tentative" fix. Does that mean anything besides that the PR isn't accepted yet?

It means I was able to quickly identify the issue and throw together a simple fix, but not to test it ;) I still don't have a working Psi4 compile environment due to the libint2 incompatibility. Once that gets fixed, I'll have to push to get newer versions of Psi4 to compile again in Fedora.

**Comment by susilehtola** [comment]
So... looks like it works; DIIS just takes it to the wrong solution
```
                          Total Energy        Delta E     RMS |[F,P]|

   @DF-UHF iter SAD:    -0.70872378254081   -7.08724e-01   0.00000e+00 
  Mixing alpha HOMO/LUMO orbitals (1,2)

   @DF-UHF iter   1:    -0.82640393434897   -1.17680e-01   1.12200e-03 DIIS
   @DF-UHF iter   2:    -0.97800267490794   -1.51599e-01   1.18746e-02 DIIS
   @DF-UHF iter   3:    -0.82673929652084    1.51263e-01   9.92192e-04 DIIS
   @DF-UHF iter   4:    -0.82768446534307   -9.45169e-04   2.11271e-03 DIIS
   @DF-UHF iter   5:    -0.82648408142877    1.20038e-03   9.77585e-06 DIIS
   @DF-UHF iter   6:    -0.82648447847074   -3.97042e-07   3.83550e-05 DIIS
   @DF-UHF iter   7:    -0.82648407869393    3.99777e-07   1.24969e-06 DIIS
   @DF-UHF iter   8:    -0.82648407827585    4.18083e-10   7.42903e-08 DIIS
   @DF-UHF iter   9:    -0.82648407827446    1.38667e-12   8.54867e-10 DIIS
   @DF-UHF iter  10:    -0.82648407827446    7.77156e-16   1.13025e-12 DIIS
  Energy and wave function converged.
```

**Comment by susilehtola** [comment]
OK, tests appear to pass now with all guesses.

I can't generate `output.ref` so I would appreciate if someone can compile the branch locally and contribute the missing files.

**Comment by JonathonMisiewicz** [comment]
Running the same test but with different options is a perfect reason to write a pytest rather than a C-test, and no `output.ref` are needed in that case. If you can't even compile Psi, I can write the Pytest myself and submit a PR to your branch, because at least I can test it easily.

Thanks for the debugging thus far!

File: psi4/src/psi4/libscf_solver/uhf.cc [hunk 1]
```diff
@@ -287,15 +289,31 @@ void UHF::form_C(double shift) {
...
-        } else {
+    if (options_.get_bool("GUESS_MIX") && !mix_performed_) {
+        if (Ca_->nirrep() != 1) {
+            outfile->Printf("  Mixing alpha HOMO/LUMO orbitals (%d,%d)\n", nalpha_, nalpha_ + 1);
             throw InputException("Warning: cannot mix alpha HOMO/LUMO orbitals. Run in C1 symmetry.",
                                  "to 'symmetry c1'", __FILE__, __LINE__);
         }
+
+        // SAD doesn't have orbitals in iteration 0, other guesses do
+        bool have_orbitals = !sad_ || (sad_ && iteration_ > 0);
```
Comment (JonathonMisiewicz) [inline psi4/src/psi4/libscf_solver/uhf.cc:300]: In my ideal world, we could just check if `Ca_` was `nullptr`, but getting Psi to such a state is beyond-the-scope of this PR.

**Comment by JonathonMisiewicz** [comment]
Accidentally pushed rather than made a PR to your fork. Oops. LMK if you object, @susilehtola.

**Review by maxscheurer — APPROVED** [review]

File: psi4/src/psi4/libscf_solver/uhf.cc (hunk 1 above)
Comment (susilehtola) [reply psi4/src/psi4/libscf_solver/uhf.cc:300]: Yup...

File: tests/pytests/test_scf_options.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        H
+        H 1 3.0
+        symmetry c1
+        """)
+
+    psi4.set_options({"reference": "uhf", "e_convergence": 12, "basis": "cc-pvdz"})
+    psi4.set_options(inp.get("options", {}))
+
+    thisSCF = psi4.energy("scf")
+    psi4.set_options(inp.get("options_late", {}))
```
Comment (susilehtola) [inline tests/pytests/test_scf_options.py:34]: @JonathonMisiewicz shouldn't this be `'late_options'` as on line 11 instead of `options_late`?

**Comment by susilehtola** [comment]
> Accidentally pushed rather than made a PR to your fork. Oops. LMK if you object, @susilehtola.

that's fine

File: tests/pytests/test_scf_options.py (hunk 2 above)
Comment (JonathonMisiewicz) [reply tests/pytests/test_scf_options.py:34]: Good call - they should be consistent.

Comment (susilehtola) [reply tests/pytests/test_scf_options.py]: Fixed

**Review by JonathonMisiewicz — APPROVED** [review]

**Review by jturney — APPROVED** [review]

**Pull request merged**